ddl/ddl_api.go

```diff
@@ -636,6 +637,113 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
 	return nil
 }
 
+func checkCreatePartitionNameUnique(part *model.PartitionInfo) error {
+	if part.Type == model.PartitionTypeRange {
+		newPars := part.Definitions
+		set := make(map[string]struct{})
+		for _, newPar := range newPars {
+			if _, ok := set[strings.ToLower(newPar.Name)]; ok {
+				return ErrSameNamePartition.GenByArgs(newPar.Name)
+			}
+			set[strings.ToLower(newPar.Name)] = struct{}{}
+		}
+	}
+	return nil
+}
+
+// checkCreatePartitionValue checks 'less than value' must be strictly increasing for each partition.
+func checkCreatePartitionValue(part *model.PartitionInfo) error {
+	if part.Type == model.PartitionTypeRange {
+		newDefs := part.Definitions
```

Comment:
Well, I mean, how about this?  If `maxvalue` exists, it must be the last one:

```
defs := part.Definitions
if strings.EqualFold(defs[len(defs) - 1], "MAXVALUE") {
	defs = defs[:len(defs) - 1]
}

for i:=1; i<len(defs); i++ {
	if strings.EqualFold(defs[i].LessThan[0], "MAXVALUE") {
		errors.Trace(ErrPartitionMaxvalue)
	}
	curr := strconv.Atoi(defs[i].LessThan[0])
	prev := strconv.Atoi(defs[i-1].LessThan[0])
	if curr > prev {
		return errors.Trace(ErrRangeNotIncreasing)
	}
}
```

Reply:
If `maxvalue` is in the middle, `prev := strconv.Atoi(defs[i-1].LessThan[0])` will fail to convert.

Reply:
For example:
```
create table employees (
	id int not null,
	hired date not null
	)
	partition by range( year(hired) ) (
		partition p1 values less than (1998),
		partition p2 values less than maxvalue,
		partition p3 values less than (2001)
	);
```

Reply:
@ciscoxll Can we use two loops:
1. the first loop is to find whether "MAXVALUE" is the last value,
2. if "MAXVALUE" is not found, we use the second loop to compare the successive digits.